File: src/tool/DevilDaggersInfo.Tool.DistributeApp/AppBuilder.fs
module DevilDaggersInfo.Tool.DistributeApp.AppBuilder

open System.Diagnostics
open System.Text
open DevilDaggersInfo.Types.Web

let getPublishCommandProperties (publishDirectoryName, runtimeIdentifier, publishMethod, singleFile) : Map<string, string> =
    let singleFileStr = singleFile.ToString()
    Map.empty
        .Add("PublishSingleFile", singleFileStr)
        .Add("SelfContained", singleFileStr)
        .Add("PublishTrimmed", "True")
        .Add("EnableCompressionInSingleFile", "True")
        .Add("PublishReadyToRun", "False")
        .Add("PublishProtocol", "FileSystem")
        .Add("TargetFramework", "net7.0")
        .Add("RuntimeIdentifier", runtimeIdentifier)
        .Add("Platform", "x64")
        .Add("Configuration", "Release")
        .Add("PublishMethod", publishMethod)
        .Add("PublishDir", publishDirectoryName)

let getPublishCommand projectFilePath publishDirectoryName toolBuildType toolPublishMethod : string =
    let runtimeIdentifier = match toolBuildType with
                            | ToolBuildType.LinuxWarp -> "linux-x64"
                            | ToolBuildType.WindowsWarp -> "win-x64"
                            | _ -> failwith "Build type not supported"
    let publishMethod = match toolPublishMethod with
                        | ToolPublishMethod.SelfContained -> "SELF_CONTAINED"
                        | _ -> "" // TODO: OK?

    let sb = StringBuilder()
    getPublishCommandProperties(publishDirectoryName, runtimeIdentifier, publishMethod, toolPublishMethod = ToolPublishMethod.SelfContained) |> Map.iter (fun key value -> sb.Append($" -p:{key}={value}") |> ignore) 
    $"publish {projectFilePath}{sb.ToString()}"

let build projectFilePath publishDirectoryName toolBuildType toolPublishMethod =
    let dotnetPublishProc = Process.Start(ProcessStartInfo(FileName = "dotnet", Arguments = getPublishCommand projectFilePath publishDirectoryName toolBuildType toolPublishMethod))
    dotnetPublishProc.WaitForExit()

    match dotnetPublishProc.ExitCode with
    | 0 -> ()
    | _ -> failwith "Could not build the app"

File: src/tool/DevilDaggersInfo.Tool.DistributeApp/AppBuilder.cs
using CliWrap;
using DevilDaggersInfo.Api.Admin.Tools;
using System.Text;

namespace DevilDaggersInfo.Tool.DistributeApp;

public static class AppBuilder
{
	public static async Task BuildAsync(string projectFilePath, string publishDirectoryName, ToolBuildType toolBuildType, ToolPublishMethod toolPublishMethod)
	{
		await Cli.Wrap("dotnet")
			.WithArguments(GetArguments(projectFilePath, publishDirectoryName, toolBuildType, toolPublishMethod))
			.ExecuteAsync();
	}

	private static string GetArguments(string projectFilePath, string publishDirectoryName, ToolBuildType toolBuildType, ToolPublishMethod toolPublishMethod)
	{
		string runtimeIdentifier = toolBuildType switch
		{
			ToolBuildType.LinuxWarp or ToolBuildType.LinuxConsole => "linux-x64",
			ToolBuildType.WindowsWarp or ToolBuildType.WindowsConsole => "win-x64",
			_ => throw new NotImplementedException(),
		};

		Dictionary<string, string> dictionary = new()
		{
			{ "PublishTrimmed", "True" },
			{ "EnableCompressionInSingleFile", "True" },
			{ "PublishReadyToRun", "False" },
			{ "PublishProtocol", "FileSystem" },
			{ "TargetFramework", "net7.0" },
			{ "RuntimeIdentifier", runtimeIdentifier },
			{ "Platform", "x64" },
			{ "Configuration", "Release" },
			{ "PublishDir", publishDirectoryName },
		};

		switch (toolPublishMethod)
		{
			case ToolPublishMethod.SelfContained:
				dictionary.Add("PublishSingleFile", "True");
				dictionary.Add("SelfContained", "True");
				dictionary.Add("PublishMethod", "SELF_CONTAINED");
				break;
			case ToolPublishMethod.Aot:
				dictionary.Add("PublishAot", "True");
				break;
			case ToolPublishMethod.Default: throw new NotImplementedException();
			default: throw new ArgumentOutOfRangeException(nameof(toolPublishMethod), toolPublishMethod, null);
		}

		StringBuilder stringBuilder = new($"publish {projectFilePath}");
		foreach (KeyValuePair<string, string> keyValuePair in dictionary)
			stringBuilder.Append(" -p:").Append(keyValuePair.Key).Append('=').Append(keyValuePair.Value);

		return stringBuilder.ToString();
	}
}
